sphinxext/icon.py

```diff
@@ -1,121 +1,322 @@
 import re
-from docutils import nodes
-from docutils.parsers.rst import roles
+import fett
+from docutils import nodes, statemachine
+from docutils.parsers.rst import roles, Directive
 from sphinx.util.nodes import split_explicit_title
+from docutils.utils.error_reporting import ErrorString
 
 PAT_STRIP = re.compile(r'-(?:o|(?:alt))$')
+ALL_SUFFIXES = ('-add', '-edit', '-exclamation', '-remove', '-restart')
+HELP_TEMPLATE = fett.Template('''
+.. list-table::
 
-# A pre-approved list of icons. Approval required to add to this list.
+{{ for name in icons }}
+   * - ``{{ name }}``
+     - :{{ role }}:`{{ name }}`
+{{ end }}
+''')
+
+# Each of the below classes describes an icon font. Each must provide:
+# * A name for the docutils icon node
+# * A name for the docutils role
+# * A prefix for the CSS class attached to the output <span>
+# * A set  of icon names
+# Approval required to add to icon lists.
 # Make sure you update the documentation in github.com:mongodb/docs-meta
 # as well when updating.
-ICONS = set([
-    'caret-right',
-    'caret-square-left',
-    'caret-square-right',
-    'check-circle',
-    'check-square',
-    'copy',
-    'edit',
-    'exclamation-circle',
-    'eye',
-    'globe',
-    'lock',
-    'pencil-alt',
-    'save',
-    'star',
-    'sync-alt',
-    'trash',
-    'trash-alt',
-    'users'
-])
-
-# Discouraged, but not prohibited (yet)
-ICONS.add('arrow-right')
-ICONS.add('asterisk')
-ICONS.add('check')
-ICONS.add('ellipsis-h')
-ICONS.add('minus')
-ICONS.add('plus')
-
-ICONSB = set([
-    'windows',
-])
+
+
+class IconsFA5:
+    """FontAwesome 5 Solid icon font"""
+    NODE_NAME = 'IconFA5'
+    ROLE_NAME = 'icon'
+    CSS_PREFIX = 'fas fa-'
+    ICONS = set([
```

Comment:
We have to define a set of icons for error-checking purposes; prior to defining the set of legal icons last year, there were several instances of nonexistent icons being used.

I'd be open to allowing any valid FontAwesome 4 & 5 icons. @skerschb any objections to this?

Reply:
Ah, that makes sense. I would like to find a solution that does not have other teams going back to the well to revise, etc.

Reply:
It seems to me to be fairly trivial to get new icons added now that we are supporting multiple font awesome versions..and the downside of all of them being allowed is that once the cats out of the bag with using whatever icon you choose it becomes very hard to standardize.

Reply:
Adding icons is absolutely trivial, but adding the whole shebang isn't a technical problem.

My thinking is that icon names are namespaced, and so handling updates should be handled by the writers. They're also better equipped for this since they know the content they're working with and what icons actually need to be rendered.

Reply:
agree, see above.

Reply:
@tony I've added the icons you requested to the allowed list. At any point, the CET team can file a ticket with us to request more; it's trivial and the turnaround time should be very fast.

If this process becomes problematic, we can and should definitely re-evaluate; we have options!